File: native/android/app/src/main/java/fr/petitejerusalem/app/PjWidgetsPlugin.java
package fr.petitejerusalem.app;

import android.content.Context;
import android.content.SharedPreferences;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Pont app → widgets d'écran d'accueil.
 *
 * La webview pousse ici les payloads JSON pré-calculés (src/services/
 * widgetService.ts) ; ils sont stockés en SharedPreferences puis les widgets
 * sont rafraîchis. Les providers lisent ce même store, le natif n'a aucune
 * logique métier, tout (horaires, libellés localisés) vient de l'app.
 *
 * Fichier versionné dans native/android/, copié dans android/ (git-ignoré)
 * par scripts/setup-android.mjs et enregistré dans MainActivity.
 */
@CapacitorPlugin(name = "PjWidgets")
public class PjWidgetsPlugin extends Plugin {
    static final String STORE = "pj_widgets";
    static final String KEY_ZMANIM = "zmanim";
    static final String KEY_DAILY = "daily";
    /** Titres des raccourcis vers la bibliothèque (le sidour, les Tehilim). */
    static final String KEY_LIBRARY = "library";

    static SharedPreferences store(Context context) {
        return context.getSharedPreferences(STORE, Context.MODE_PRIVATE);
    }

    @PluginMethod
    public void setPayloads(PluginCall call) {
        Context context = getContext();
        SharedPreferences.Editor editor = store(context).edit();
        String zmanim = call.getString("zmanim");
        if (zmanim != null) editor.putString(KEY_ZMANIM, zmanim);
        String daily = call.getString("daily");
        if (daily != null) editor.putString(KEY_DAILY, daily);
        String library = call.getString("library");
        if (library != null) editor.putString(KEY_LIBRARY, library);
        editor.apply();

        // Seuls les widgets dont le payload a changé sont redessinés.
        if (zmanim != null) {
            PjWidgetProvider.requestUpdate(context, HorairesWidgetProvider.class);
            PjWidgetProvider.requestUpdate(context, EssentialsWidgetProvider.class);
            PjWidgetProvider.requestUpdate(context, ZmanimShortcutWidgetProvider.class);
        }
        if (daily != null) {
            PjWidgetProvider.requestUpdate(context, LectureWidgetProvider.class);
            PjWidgetProvider.requestUpdate(context, LectureShortcutWidgetProvider.class);
        }
        if (library != null) {
            PjWidgetProvider.requestUpdate(context, LibraryWidgetProvider.class);
            PjWidgetProvider.requestUpdate(context, SidourWidgetProvider.class);
            PjWidgetProvider.requestUpdate(context, TehilimWidgetProvider.class);
        }
        call.resolve();
    }
}

File: native/ios/App/PjWidgetsPlugin.swift
import Capacitor
import Foundation
import WidgetKit

/**
 * Pont app → widgets d'écran d'accueil (pendant iOS de
 * native/android/.../PjWidgetsPlugin.java).
 *
 * La webview pousse ici les payloads JSON pré-calculés
 * (src/services/widgetService.ts) ; ils sont stockés dans l'App Group, seul
 * espace lisible par l'extension de widgets, puis WidgetKit recharge les
 * timelines.
 *
 * Fichier copié dans la cible App du projet Xcode (généré, non versionné) par
 * scripts/setup-ios.mjs, qui l'y enregistre aussi via PjViewController, voir
 * docs/app-widgets.md.
 */
@objc(PjWidgetsPlugin)
public class PjWidgetsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PjWidgetsPlugin"
    public let jsName = "PjWidgets"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setPayloads", returnType: CAPPluginReturnPromise)
    ]

    /// Doit exister dans Signing & Capabilities des DEUX cibles (App + widgets).
    static let appGroup = "group.fr.petitejerusalem.app"

    @objc func setPayloads(_ call: CAPPluginCall) {
        guard let defaults = UserDefaults(suiteName: Self.appGroup) else {
            call.reject("App Group \(Self.appGroup) indisponible, vérifier la capability App Groups.")
            return
        }
        // Seuls les widgets dont le payload a changé sont rechargés : chaque
        // reload consomme le budget de rafraîchissement WidgetKit.
        if let zmanim = call.getString("zmanim") {
            defaults.set(zmanim, forKey: "zmanim")
            for kind in ["HorairesWidget", "EssentialsWidget", "ZmanimShortcutWidget"] {
                WidgetCenter.shared.reloadTimelines(ofKind: kind)
            }
        }
        if let daily = call.getString("daily") {
            defaults.set(daily, forKey: "daily")
            WidgetCenter.shared.reloadTimelines(ofKind: "LectureWidget")
            WidgetCenter.shared.reloadTimelines(ofKind: "LectureShortcutWidget")
        }
        // Bibliothèque : les libellés des raccourcis. Ils ne bougent qu'avec
        // la langue, l'app ne les renvoie donc presque jamais.
        if let library = call.getString("library") {
            defaults.set(library, forKey: "library")
            for kind in ["LibraryWidget", "SidourWidget", "TehilimWidget"] {
                WidgetCenter.shared.reloadTimelines(ofKind: kind)
            }
        }
        call.resolve()
    }
}
